triton_kernel_agent/worker.py

```diff
@@ -218,37 +217,128 @@ def _detect_pytorch_compute(self, kernel_code: str) -> Optional[str]:
+                code = f.read()
+
+            exec_globals = {
+                "__name__": "__main__",
+                "__file__": str(test_file),
+            }
+            exec(code, exec_globals)
+
+            # Template enforces testsfiles to call sys.exit(0)
+            # This should NOT be reached
+            result_queue.put(
+                (
+                    False,
+                    stdout_buffer.getvalue(),
+                    "Test misformatted; did not call sys.exit(#) "
+                    + stderr_buffer.getvalue(),
+                )
             )
 
-            success = result.returncode == 0
-            self.logger.info(
-                f"Test {'passed' if success else 'failed'} with code {result.returncode}"
+        except SystemExit as e:
+            # Test code template calls sys.exit()
+            if e.code == 0 or e.code is None:
+                result_queue.put(
+                    (True, stdout_buffer.getvalue(), stderr_buffer.getvalue())
+                )
+            else:
+                # Non-zero exit code means failure
+                result_queue.put(
+                    (
+                        False,
+                        stdout_buffer.getvalue(),
+                        f"Test exited with code {e.code} " + stderr_buffer.getvalue(),
+                    )
+                )
+
+        except BaseException:
+            result_queue.put(
+                (
+                    False,
+                    stdout_buffer.getvalue(),
+                    stderr_buffer.getvalue() + traceback.format_exc(),
+                )
             )
 
-            return success, result.stdout, result.stderr
+        finally:
+            # Restore original stdout/stderr
+            sys.stdout = original_stdout
+            sys.stderr = original_stderr
+
+    def _run_test(self) -> Tuple[bool, str, str]:
+        """
+        Run the test script and capture results using multiprocessing.
+
+        Returns:
+            Tuple of (success, stdout, stderr)
+        """
+        # Create process to run the test
+        result_queue = mp.Queue()
+        process = mp.Process(
+            target=self._run_test_process,
+            args=(self.test_file, self.workdir, result_queue),
+        )
+        process.start()
+        process.join(timeout=30)
+
+        # Check if process is still alive (timeout)
+        if process.is_alive():
+            process.terminate()
+            process.join(timeout=5)
+            if process.is_alive():
+                process.kill()
+                process.join()
 
-        except subprocess.TimeoutExpired:
             self.logger.error("Test timed out")
             return False, "", "Test execution timed out after 30 seconds"
-        except Exception as e:
-            self.logger.error(f"Test execution error: {e}")
-            return False, "", str(e)
+
+        # Get result from queue
+        if not result_queue.empty():
```

Comment:
Relying on result_queue.empty() to decide whether a result is available is fragile: Queue.empty() isn’t guaranteed to be accurate across processes, so we could return “ended without result” even if the child actually put something on the queue.